clients/lsp-javascript.el

```diff
@@ -790,20 +790,42 @@ name (e.g. `data' variable passed as `data' parameter)."
   (when-let ((workspace (lsp-find-workspace 'ts-ls (buffer-file-name))))
     (eq 'initialized (lsp--workspace-status workspace))))
 
-(defun lsp-clients-typescript-project-ts-server-path ()
-  "Return the project local TS server path."
-  (f-join (lsp-workspace-root) "node_modules" "typescript" "lib" "tsserver.js"))
+(defun lsp-clients-typescript-require-resolve (&optional dir)
+  "Get the location of the typescript.
+Use Node.js require.
+The node_modules directory structure is suspect
+and should be trusted as little as possible.
+If you call require in Node.js,
+it should take into account the various hooks.
+For example, yarn PnP.
+
+Optional argument DIR specifies the working directory
+to run the command in."
+  (let* ((default-directory (or dir default-directory))
+         (output
+          (string-trim-right
+           (shell-command-to-string
+            "node -e \"console.log(require.resolve('typescript'))\""))))
```

Comment:
Normally, we should wrap this with `shell-quote-argument`. See https://github.com/emacs-grammarly/keytar/blob/69eb60440c21879ce65c7205f070d59574320bb4/keytar.el#L59-L71.

Not sure why we don't have something like this under the global `lsp-` namespace since it's very useful. 🤔

Reply:
I was certainly concerned about going through the shell as well and would have preferred not to if possible.
However, I could not immediately figure out how to execute the command directly and receive it as a string without creating a separate buffer or something. In this case, the argument is a fixed string constant, and since it is not passed through an argument, etc., I thought it would not be a security issue, so I passed it as it is.

Reply:
Make sense to me.